### ISSUE TITLE
Can't Connect via Socket

### ISSUE THREAD

**Issue by pelmered** [Sequel-Ace/Sequel-Ace#113]
- Sequel Ace Version: 2.0.2
- macOS Version: 10.15.5 (Catalina)
- MySQL Version: MariaDB 10.4.13

**Description**
Sequel Ace can't connect to local MariaDB though socket with empty password. 
<img width="435" alt="Screenshot 2020-06-26 at 11 57 08" src="https://user-images.githubusercontent.com/680058/85845256-40ecc600-b7a4-11ea-9a5d-4dad6dfbe39e.png">

**Steps To Reproduce**
1. Install MariaDB with homebrew `brew reinstall mariadb`
2. Setup connection with socket, your mac user username and empty password (the default credentials when installing through brew)
<img width="533" alt="Screenshot 2020-06-26 at 11 56 55" src="https://user-images.githubusercontent.com/680058/85845197-316d7d00-b7a4-11ea-84cf-e436d1b2b072.png">
3. Try to connect

**Expected Behaviour**
It should connect successfully.

**Additional Context**
- The exact same connection settings work fine in Sequel Pro

- I have also tried to specify the socket path: `/tmp/mysql.sock`

- My guess is that this has something to do with the empty password. 

- Connecting though CLI works fine as well. Both these are working:
`mysql`
`mysql --user=peter --socket=/tmp/mysql.sock`

**Comment by gboudreau** [comment]
I can confirm socket connections are failing...
I can connect using 127.0.0.1, user root, no password. But I can't connect using /tmp/mysql.sock, user root.

**Comment by Jason-Morcos** [comment]
We probably can't access the tmp dir maybe? Cause of sandbox. Sockets not working for me either.

**Comment by gboudreau** [comment]
Indeed, we might have to use a file selector to select the socket, and use a bookmark... That would mean making that field mandatory.

**Comment by Jason-Morcos** [comment]
I think that's probably the right call, @gboudreau. Gonna take a look at the new SSL/SSH keys bookmarks PR (James tweaked it last night) and merge that if good - perhaps we could leverage the same stuff that does, just adding a file selector for socket?

**Comment by Jason-Morcos** [comment]
Other PR merged if we want to leverage similar capabilities! https://github.com/Sequel-Ace/Sequel-Ace/pull/105

**Comment by jamesstout** [comment]
> Indeed, we might have to use a file selector to select the socket, and use a bookmark... That would mean making that field mandatory.

I can't get the NSOpenPanel to display the socket file... it's there on the filesystem, not shown in the panel:

![Screenshot 2020-06-27 at 10 18 01 PM](https://user-images.githubusercontent.com/1576190/85924551-b2706580-b8c5-11ea-9b99-231eeea67ed5.png)
![Screenshot 2020-06-27 at 10 30 07 PM](https://user-images.githubusercontent.com/1576190/85924581-dc298c80-b8c5-11ea-98a1-1dcaa343ab96.png)

**Comment by Jason-Morcos** [comment]
> > Indeed, we might have to use a file selector to select the socket, and use a bookmark... That would mean making that field mandatory.
> 
> 
> 
> I can't get the NSOpenPanel to display the socket file... it's there on the filesystem, not shown in the panel:
> 
> 
> 
> ![Screenshot 2020-06-27 at 10 18 01 PM](https://user-images.githubusercontent.com/1576190/85924551-b2706580-b8c5-11ea-9b99-231eeea67ed5.png)
> 
> ![Screenshot 2020-06-27 at 10 30 07 PM](https://user-images.githubusercontent.com/1576190/85924581-dc298c80-b8c5-11ea-98a1-1dcaa343ab96.png)
> 
> 

@jamesstout is that with show hidden files/folders enabled in the open panel?

**Comment by jamesstout** [comment]
> @jamesstout is that with show hidden files/folders enabled in the open panel?

Yep.

**Comment by jamesstout** [comment]
If you set the app to be not sandboxed, you can connect, but the socket file still doesn't appear in the open panel.

![Screenshot 2020-06-28 at 5 14 52 PM](https://user-images.githubusercontent.com/1576190/85943588-94594280-b963-11ea-88da-c771bc357219.png)

**Comment by Jason-Morcos** [comment]
Does it work if we require the user to give us access to the whole tmp directory via security scoped bookmark? Like if to use a socket you must go select the tmp directory and grant us access via file browser? Idk if it would get approved, but we could make it so it really only popped up if you tried to use socket and was indicated as an opt-in kind of thing?

**Comment by jamesstout** [comment]
I don't think so, pretty sure I tested that. The sockets do appear if you use 

`[urlArr setArray:[fm contentsOfDirectoryAtPath:selectedFilePath error:&err]];`

`urlArr` contains the socket. But it hasn't been selected by the user. I'll try again later. Working hours here now.

**Comment by jamesstout** [comment]
Yeah, doesn't work. You can see here I have` /tmp `(actually `/private/tmp`) in the bookmarks, but when I try to use `/tmp/mysqld.sock`, I get the error reported above.

![Screenshot 2020-06-29 at 3 56 13 PM](https://user-images.githubusercontent.com/1576190/85988681-fd06f480-ba21-11ea-9db4-9d437c2287bb.png)

**Comment by brandondove** [comment]
I'm also having a similar problem connecting via socket. It's not te exact same issue, so I'm not sure if I should be posting this here or as a new issue.

- Sequel Ace Version: 2.1.0, build 2042
- macOS Version: 10.15.5 (Catalina)
- MySQL Version: MySQL 8.0.16
- Using [Local by Flywheel](https://localwp.com)

Description
Sequel Ace can't connect to MySQL using socket located at `/Users/{USERNAME}/Library/Application Support/Local/run/{SITEID}/mysql/mysqld.sock`

![Screen_Shot_2020-06-30_at_5_08_46_PM](https://user-images.githubusercontent.com/300348/86189299-0a8dae80-baf6-11ea-9d1a-d7fcdbfaa8b7.png)

Steps To Reproduce

1. Install Local by Flywheel
2.  Create custom site
3.  Turn on site
4.  Setup connection with socket, user and password are both root - 
![Screen Shot 2020-06-30 at 5 12 19 PM](https://user-images.githubusercontent.com/300348/86189305-124d5300-baf6-11ea-8a7a-a19c008ee304.png)
5. Try to connect

Expected Behaviour
It should connect successfully.

Additional Context

The exact same connection settings work fine in Sequel Pro thought the tables don't render properly once connected. I have two theories about why this isn't working.

1. The socket and database are located in a user directory rather than a directory that is accessible system-wide.
2. The full path has a space in it between "Application" and "Support". I've tried adding a backslash before the space to escape it, but that didn't seem to help at all.

**Comment by Jason-Morcos** [comment]
@brandondove Yup it's the same issue. The issue is that Sequel Ace is sandboxed and can't access _any_ socket. Nothing to do with your specific setup. We're trying to find a workaround, but the app has to be sandboxed to be available on the App Store.

**Comment by jb510** [comment]
> We're trying to find a workaround, but the app has to be sandboxed to be available on the App Store.

TY.  Incidentally, I figured that and tried granting Ace "Full Disk Access" in the Security & Privacy panel.  Still wouldn't access the socket at ~/Library/Application Support/Local Beta/run/swgVOs0cS/mysql/mysqld.sock

**Comment by CeoTang** [comment]
![Screenshot 2020-07-08 at 11 41 02 AM](https://user-images.githubusercontent.com/14979633/86872952-eb070f00-c10f-11ea-97fb-0c05496d8320.png)
How can i fix this problem?

**Comment by njvack** [comment]
I suspect you could skip the sandboxing and just talk to the socket without any fuss (this was a very confusing experience for me!) by doing a non-MAS release (see also #58).

**Comment by jamesstout** [comment]
> > Indeed, we might have to use a file selector to select the socket, and use a bookmark... That would mean making that field mandatory.
> 
> I can't get the NSOpenPanel to display the socket file... it's there on the filesystem, not shown in the panel:
> 

Quinn (The Eskimo) [replied](https://developer.apple.com/forums/thread/124895?answerId=624312022#624312022) to my query: _"Can NSOpenPanel ever show a socket? e.g. /tmp/mysql.sock?"_

> Doesn’t look like it. It makes sense that these are hidden because the sandbox extension it would issue to you app wouldn’t be helpful (because file sandbox extensions are for file operations, not for networking).

**Comment by GarrettAlbright** [comment]
This is kind of a bummer. I'd really like to finally have Sequel Pro in dark mode but I'd rather stick to just socket networking if possible.

Here's a workaround for MySQL installed with MacPorts.

1. Figure out what "MySQL" variant and version you actually have installed and running. In my case I have `mariadb-10.1`.
```cli
> port select --show mysql
The currently selected version for 'mysql' is 'mariadb-10.1'.
```
2. Edit the corresponding `my.conf` file at `/opt/local/etc/{variant-version}/my.cnf`. You will need root privileges (sudo).
3. Add the following:
```
[mysqld]
skip_networking=0
bind_address=127.0.0.1
```
4. Restart the server.
```cli
> sudo port reload {variant-version}-server
```
5. Set up a TCP/IP connection in Sequel Ace with Host set to "127.0.0.1". Other settings (excluding "Port") should be the same as what you had for your socket connection.

Note that if you want toi have multiple versions of MySQL running on your system concurrently, you'll need to add `port={some number}` with a different number for each version in step 3, and set the port number for each version in step 5.

Here's hoping Sequel Ace can get support for this worked out in the future. If it can't be done in the MAS sandbox, is a non-MAS release on the table?

**Comment by Jason-Morcos** [comment]
What's the benefit of using a socket exactly? Why not connect via TCP/IP? I know Sequel Pro was considering ditching socket pro support completely (as we may be forced to do as well).

Additionally, I have no desire to distribute both a sandboxed and non-sandboxed version of the app. That introduces a slew of issues for longterm support (supporting basically two apps instead of one). However, you can download the source code for Sequel Ace, disable the sandbox  in Xcode, and build a custom build if you desire. I just think it's not sustainable long-term to support two different versions of the same app.

**Comment by GarrettAlbright** [comment]
> What's the benefit of using a socket exactly? Why not connect via TCP/IP? I know Sequel Pro was considering ditching socket pro support completely (as we may be forced to do as well).

It allows us to avoid the security concerns of running yet another theoretically network-accessible server, and the connection to the server is faster and has less overhead since requests go through local IPC rather than going through the network stack.

Granted, on local development machines, which I'm guessing is how most people use Sequel Pro/Ace, these things aren't as big of a concern, but they're a concern nonetheless.

> Additionally, I have no desire to distribute both a sandboxed and non-sandboxed version of the app. That introduces a slew of issues for longterm support (supporting basically two apps instead of one). However, you can download the source code for Sequel Ace, disable the sandbox in Xcode, and build a custom build if you desire. I just think it's not sustainable long-term to support two different versions of the same app.

Okay, fair enough. I get it. I hope a solution to the socket issue presents itself, then.

For what it's worth, the built-in MySQL client in Coda 2, which is a MAS app, does support socket connections, so there must be a way.

**Comment by jamesstout** [comment]
> For what it's worth, the built-in MySQL client in Coda 2, which is a MAS app, does support socket connections, so there must be a way.

I have Coda 2, will have a look. However, Coda 2 is no longer on the MAS: [Coda 2.5 and the Mac App Store](https://panic.com/blog/coda-2-5-and-the-mac-app-store/).

**Comment by CaddyDz** [comment]
I ended up using TCP connection
if anyone has trouble with that, change `/usr/local/etc/my.cnf` 
```
[mysqld]
skip-grant-tables
#
# This group is read both both by the client and the server
# use it for options that affect everything
#
[client-server]

#
# include all files from the config directory
#
!includedir /usr/local/etc/my.cnf.d
```
⚠️ This disables all security ⚠️ 
But TCP connection to `127.0.0.1` will now work

I wish there was an unsandboxed release of Sequel Ace

**Comment by jamesstout** [comment]
> I have Coda 2, will have a look.

Ah, my version is the non-appstore version.

**Comment by ShivamJoker** [comment]
Any fix for this guys I am not able to use the tcp/ip also

**Comment by jamesstout** [comment]
Hah, look what's inside the Coda 2 Resources folder:

![Screenshot 2020-08-03 at 5 08 01 PM](https://user-images.githubusercontent.com/1576190/89166350-8030e280-d5ac-11ea-8bb6-c77cfcc568c6.png)

**Comment by jamesstout** [comment]
> Any fix for this guys I am not able to use the tcp/ip also

Trying...

**Comment by GarrettAlbright** [comment]
> I have Coda 2, will have a look. However, Coda 2 is no longer on the MAS: [Coda 2.5 and the Mac App Store](https://panic.com/blog/coda-2-5-and-the-mac-app-store/).

Ah, my mistake. I had forgotten they moved off of the MAS way back when.

**Comment by jamesstout** [comment]
OK, I've got it working, with one big caveat...

You need to change `/usr/local/etc/my.cnf` and add:

```yaml
[mysqld]
socket=/Users/YourUser/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock
```

Is that no good?

**Comment by GarrettAlbright** [comment]
Does a symlink suffice?

**Comment by jamesstout** [comment]
alas no

**Comment by stychos** [comment]
Apple definitely needs to do something with sockets. Just thought what if mysqld will come sandboxed 🤣

**Comment by jamesstout** [comment]
> EDIT: To clarify, I mean if I make a symlink at that location which points to the original socket location, would that work?

Nope, just tried it... it finds the symlink, but when it tries to connect it fails.

```
2020-08-03 19:33:06.854573+0800 Sequel Ace[53111:6224889] checking = /Users/james/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock
2020-08-03 19:33:06.854651+0800 Sequel Ace[53111:6224889] found = /Users/james/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock
```

```
james@Jamess-iMac: ~/Library/Containers/com.sequel-ace.sequel-ace/Data
$ ll /private/tmp/mysql.sock
srwxrwxrwx  1 james  wheel  0 Aug  3 19:30 /private/tmp/mysql.sock
```

![Screenshot 2020-08-03 at 7 36 08 PM](https://user-images.githubusercontent.com/1576190/89178798-06572400-d5c1-11ea-967d-3502a8444a82.png)

**Comment by Jason-Morcos** [comment]
Until we can figure out a better sandbox workaround for sockets, I believe it may be worth adding info on this workaround into the app (if you select socket, it pops up an alert explaining this process). Does that seem like it makes sense to you too, @jamesstout?

**Comment by jamesstout** [comment]
Yeah sounds good.

**Comment by Jason-Morcos** [comment]
@Sequel-Ace/maintainers I forget who, but someone contacted Apple about this. Have we heard anything from Apple?

**Comment by jamesstout** [comment]
> @Sequel-Ace/maintainers I forget who, but someone contacted Apple about this. Have we heard anything from Apple?

I did but got nothing back.

**Comment by jamesstout** [comment]
> Until we can figure out a better sandbox workaround for sockets, I believe it may be worth adding info on this workaround into the app (if you select socket, it pops up an alert explaining this process). Does that seem like it makes sense to you too, @jamesstout?

Done in #408 Please take a look @Jason-Morcos

**Comment by CeoTang** [comment]
![Screenshot 2020-10-22 at 4 48 55 PM](https://user-images.githubusercontent.com/14979633/96848058-82595d80-1486-11eb-8513-b63a7fc02362.png)
It does not work on sequel ace, MySQL Workbench also fixed this problem

**Comment by Jason-Morcos** [comment]
> ![Screenshot 2020-10-22 at 4 48 55 PM](https://user-images.githubusercontent.com/14979633/96848058-82595d80-1486-11eb-8513-b63a7fc02362.png)
> It does not work on sequel ace, MySQL Workbench also fixed this problem

Have you tried the instructions in the documentation?
https://sequel-ace.com/get-started/
>Unfortunately, due to sandboxing nature, Sequel Ace is not allowed to connect to the sockets which are out of the Sandbox. As a workaround, you can create a socket in ~/Library/Containers/com.sequel-ace.sequel-ace/Data and connect to it. This can be done by putting these lines to your MySQL configuration file (usually, my.cnf):
 >>[mysqld]
 >>socket=/Users/YourUserName/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock

**Comment by CeoTang** [comment]
> > ![Screenshot 2020-10-22 at 4 48 55 PM](https://user-images.githubusercontent.com/14979633/96848058-82595d80-1486-11eb-8513-b63a7fc02362.png)
> > It does not work on sequel ace, MySQL Workbench also fixed this problem
> 
> @CeoTang Have you tried the instructions in the documentation?
> https://sequel-ace.com/get-started/
> 
> > Unfortunately, due to sandboxing nature, Sequel Ace is not allowed to connect to the sockets which are out of the Sandbox. As a workaround, you can create a socket in ~/Library/Containers/com.sequel-ace.sequel-ace/Data and connect to it. This can be done by putting these lines to your MySQL configuration file (usually, my.cnf):
> > > [mysqld]
> > > socket=/Users/YourUserName/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock

it's because I'm using MAMP PRO. When I stop the SQL service, the mysql.sock will be gone.

**Comment by Kaspik** [comment]
Just checking - is this still an issue on 3.1.1?

**Comment by noseglid** [comment]
Yes it is.
![image](https://user-images.githubusercontent.com/933880/106274915-52fc7e80-6235-11eb-9376-09d7f3b4b763.png)
![image](https://user-images.githubusercontent.com/933880/106275039-8808d100-6235-11eb-92bc-88c11ef702a1.png)

**Comment by jamesstout** [comment]
Spent some time yesterday playing with [STPrivilegedTask](https://github.com/sveinbjornt/STPrivilegedTask), and managed to write a CLI app that took a path (e..g. `/tmp/mysql.sock`) and requested and granted access to it (asking for password). It returns an `AuthorizationRef` struct. However, the app cannot be part of a sandboxed bundle. So we'd need a helper app to pass this auth back to SA to get access to the socket. I found some [very oldddd code](https://developer.apple.com/library/archive/samplecode/BetterAuthorizationSample/Introduction/Intro.html#//apple_ref/doc/uid/DTS10004207) from Apple that shows how to install a helper app with a launchd plist etc.

Actually just found some updated code that uses XPC. Will check that out.

**Comment by Jason-Morcos** [comment]
> Spent some time yesterday playing with [STPrivilegedTask](https://github.com/sveinbjornt/STPrivilegedTask), and managed to write a CLI app that took a path (e..g. `/tmp/mysql.sock`) and requested and granted access to it (asking for password). It returns an `AuthorizationRef` struct. However, the app cannot be part of a sandboxed bundle. So we'd need a helper app to pass this auth back to SA to get access to the socket. I found some [very oldddd code](https://developer.apple.com/library/archive/samplecode/BetterAuthorizationSample/Introduction/Intro.html#//apple_ref/doc/uid/DTS10004207) from Apple that shows how to install a helper app with a launchd plist etc.
> 
> 
> 
> Actually just found some updated code that uses XPC. Will check that out.

Awesome, @jamesstout!
You may also find this thread and linked code helpful - https://github.com/Sequel-Ace/Sequel-Ace/issues/346

**Comment by RodolfoGS** [comment]
I've a similar issue but connecting via SSH. `Sequel Age Version 3.2.3 Build 3024`

I tested Sequel Pro with the same config in the same machine and I can connect perfectly with Sequel Pro.

Anyone have idea that could be happening?

I keep the SSH Password empty to use my public/private keys from ~/.ssh/

<img width="455" alt="Untitled 2" src="https://user-images.githubusercontent.com/870365/110964469-d6bb8780-8318-11eb-9643-c83f888179b5.png">

<img width="281" alt="Untitled" src="https://user-images.githubusercontent.com/870365/110963799-20f03900-8318-11eb-8ee9-d8fd8e5a25e4.png">

**Comment by Jason-Morcos** [comment]
> I've a similar issue but connecting via SSH. `Sequel Age Version 3.2.3 Build 3024`
> 
> I tested Sequel Pro with the same config in the same machine and I can connect perfectly with Sequel Pro.
> 
> Anyone have idea that could be happening?
> 
> I keep the SSH Password empty to use my public/private keys from ~/.ssh/
> 
> <img alt="Untitled 2" width="455" src="https://user-images.githubusercontent.com/870365/110964469-d6bb8780-8318-11eb-9643-c83f888179b5.png">
> 
> <img alt="Untitled" width="281" src="https://user-images.githubusercontent.com/870365/110963799-20f03900-8318-11eb-8ee9-d8fd8e5a25e4.png">

This is not related to this socket issue. Please post in a new issue if you're having issues instead of attaching onto an unrelated issue :)

Sequel Pro could magically pull keys and access from your .ssh dir, but Sequel Ace cannot as Sequel Ace is sandboxed. Please either select your SSH key in the SSH Password area, or grant Ace access to your ssh config in the app's preferences _and_ grant ace access to the keys in the files section of Ace's preferences. The website may be helpful for help here - https://sequel-ace.com/get-started/ (see "Sequel Ace doesn’t read my ~/.ssh/config parameters.")

**Comment by CaddyDz** [comment]
@RodolfoGS click the key icon next to the password field and select your private ssh key

**Comment by benlieb** [comment]
Has this been solved?

**Comment by fitzyjoe** [comment]
I can't connect via socket.  There was a popup that appeared telling me to add configuration to my.cnf.  I did that and restarted and the sock file appeared in the new location.  Sequel Ace still could not connect.

**Comment by fitzyjoe** [comment]
I connected via TCP to localhost.

**Comment by al404** [comment]
MacOS Monterey 12.1
MAMP 6.6

I can connect only via TCP / IP but remember in MAMP to enable on MySQL tab "Allow network access to MySQL" on mine was disable since MySQL Pro did work via socket

**Comment by ayakaida** [comment]
Mac Monterey 12.3.1
Apple Silicon M1 , 2020
Sequel Ace Version 3.5.2
MAMP PRO 6.6.2

I had the same issue but it has been solved once I checked 'Allow network access to MySQL' (Only from this computer) in MAMP > SERVERS & SERVICES > MySQL. According to the Information tooltip,  Sequel Ace requires this setting
<img width="1272" alt="Screen Shot 2022-04-23 at 4 31 50 PM" src="https://user-images.githubusercontent.com/78943333/164877971-ea99a9e0-4b2e-457a-8a07-9cae93a69ec5.png">
<img width="1272" alt="Screen Shot 2022-04-23 at 4 31 37 PM" src="https://user-images.githubusercontent.com/78943333/164877973-51bab405-afb1-4735-8239-c22e823da428.png">

**Comment by designosis** [comment]
None of the above does anything for me. Latest MAMP PRO, latest Sequel Ace, can connect via SSH but not via SOCKET.

![Screen Shot 2022-11-20 at 1 37 43 PM](https://user-images.githubusercontent.com/146623/202925106-03f6fb83-531a-4da6-8f8f-62b21f8e98c8.png)

ANY tips would be most welcome, sockets are SO much faster for me than SSH (even locally).

**Comment by yugaego** [comment]
Summing up the current state of things according to my experience on macOS Ventura 13.4 with Sequel Ace Version 4.0.5 Build 20046 (and Macports mysql8-server 8.0.33 installation).

The [documented workaround](https://sequel-ace.com/get-started/) writing a socket file to the container

```
# sudo vim /opt/local/etc/mysql8/my.cnf
[mysqld]
 socket=/Users/YourUserName/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock
```

resulted in the following MySQL error:

```
[ERROR] [MY-010273] [Server] Could not create unix socket lock file
 /Users/user/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock.lock.
```

The error is most probably caused by **_mysql** user/group permissions not matching the container's ones. Anyway, instead of fixing permissions, I preferred to follow a solution enabling TCP/IP access [suggested earlier](https://github.com/Sequel-Ace/Sequel-Ace/issues/113#issuecomment-667606874).

So, after adding the following options ([documented here](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_skip_networking)):

```
# sudo vim /opt/local/etc/mysql8/my.cnf
[mysqld]
skip-networking=OFF
bind-address=127.0.0.1
```

and restarting `mysqld` service, connection via TCP/IP with Sequel Ace was established successfully.

Related notes:
- Providing Full Disk Access in Privacy & Security Settings to Sequel Ace [doesn't affect the behavior](https://github.com/Sequel-Ace/Sequel-Ace/issues/113#issuecomment-652168600).
- MySQLWorkbench allows to connect via default socket.

**Comment by crackwilding** [comment]
I've been struggling with this issue for two days. I've tried:

- Changing the socket to `/Users/YourUserName/Library/Containers/com.sequel-ace.sequel-ace/Data/mysql.sock`. This results in an error (`Could not create unix socket lock file`). I tried adding the `_mysql` user to the group that owns that directory to no avail.
- Changing the socket to `/tmp` as Sequel Ace seems to want. I was able to create the socket file but Sequel Ace still says it can't use it. 
- Enabling TCP/IP access. 
- I also tried changing the socket to a directory in my home folder, which I added to the Sequel Ace sandbox. No dice there either. 

Tried a million variations on each of these. I'm out of ideas. For the time being I'm going to have to fall back to PHPMyAdmin.

**Comment by Jason-Morcos** [comment]
Sockets are fully blocked by the macOS sandbox for App Store apps. You will need to connect via TCP/IP for sure

**Comment by dantheperson** [comment]
Can i suggest removing the socket option from the connection dialog until such a time as the feature works?

That would have saved me a whole lot of time troubleshooting why it didn't work.

**Comment by njvack** [comment]
I'm on macOS 15.3.1, Sequel Ace 4.2.1 and Percona Server; I'm connecting via socket and it works fine. 

I'm not saying as sockets are _good_ in sandboxed apps but they are absolutely not fully blocked by the OS.

**Comment by ronilaukkarinen** [comment]
For us TCP/IP AND Socket are not working. `sudo mysql -u root -p` or `sudo mysql` works.

**Comment by m11y** [comment]
A third workaround, for the cases Options 1 and 2 can't reach: **relay the socket into Sequel Ace's App Group container instead of moving it.**

Both existing options require control over the thing that owns the socket — you edit `my.cnf` to relocate it, or you turn on TCP. That leaves out every socket you *don't* own. The one that pushed me here is an SSH agent: I keep my SSH keys in a password manager (Bitwarden, but 1Password/Secretive/gpg-agent are the same shape), and its agent socket lives inside *that* app's container. I can't relocate it, and `IdentityAgent` only accepts a UNIX-domain socket, so "just use TCP" isn't available either.

### The failure looks like an auth problem, but isn't

Sequel Ace's tunnel assistant asked for an SSH password, then failed with "could not be authenticated by the remote host". Nothing in the app's output mentions the sandbox. The actual cause only shows in the kernel log:

```
log show --last 10m --predicate 'process == "kernel" AND eventMessage CONTAINS "Sandbox"' | grep deny
```

```
kernel: (Sandbox) Sandbox: ssh(68324) deny(1) network-outbound <path to the agent socket>
```

That matches what @Jason-Morcos described in #346 — `NSTask`s share the sandbox they're launched from, so the spawned `ssh` inherits it. Key auth never happens, and ssh quietly falls back to password auth. Worth knowing because the symptom points at credentials and the cause is somewhere else entirely.

### The workaround

Run a relay **outside** the sandbox that listens on a socket **inside** a directory Sequel Ace is allowed to reach, and forwards to the real one:

```bash
socat \
  "UNIX-LISTEN:$HOME/Library/Group Containers/<TEAM_ID>.sequel-ace/agent.sock,fork,mode=600" \
  "UNIX-CONNECT:<path to the real socket>"
```

`<TEAM_ID>` is Sequel Ace's signing team identifier — the group container matches the `com.apple.security.application-groups` entitlement in its code signature, so you can read it off the app:

```bash
codesign -d --entitlements :- "/Applications/Sequel Ace.app" | grep -A2 application-groups
```

Then point the app at the relay path. For the SSH agent case that's one line in the config file Sequel Ace reads (Preferences → Network → SSH config file):

```
Host your-host
    IdentityAgent "~/Library/Group Containers/<TEAM_ID>.sequel-ace/agent.sock"
```

For a database socket it should work the same way — set the connection's socket path to the relay instead of the real one — though I've only tested the SSH agent case myself.

### Two things that cost me time

**Quote the path.** `Group Containers` contains a space, and unquoted, ssh rejects the line with `keyword identityagent extra arguments at end of line` **and terminates on the whole file**. That breaks every host in your config, not just the one you edited — a much more confusing failure than the one you started with.

**A symlink does not work.** The sandbox evaluates the resolved path, so a link pointing out of the container is denied exactly like the original. It has to be a relay that actually re-opens the connection from an unsandboxed process.

### Why the group container

Option 1 in the summary above uses the app's own container (`~/Library/Containers/com.sequel-ace.sequel-ace/Data/`), which works too. I used the App Group container because that's what group containers are *for* — shared IPC — and because it survives anything that resets the app container. Either is inside the sandbox's allowed set; the app container is the more conservative choice if you want to match what's already documented here.

### Caveats, stated plainly

- The relay has to be running, or you're back to the original symptom with a different cause. Under `launchd` with `KeepAlive` it comes back on its own. A bare background process gets reaped eventually — mine was killed by memory pressure mid-test and the failure looked exactly like the sandbox denial it had just fixed.
- It holds no key material and opens no network port; it only copies bytes between two local UNIX sockets. But it does mean anything that can reach the group container path can reach your agent, so `mode=600` is not optional.
- This doesn't make sandboxing wrong or unnecessary — the sandbox is denying exactly the kind of delegation it exists to deny. It just has no way to express "the user wants *this* app to reach *this* agent", and there's no TCC category for it.

This is a user-side workaround, not a request to change the app. Posting it because `socat`, `relay` and `group container` don't appear anywhere in this thread, and the SSH-agent variant isn't covered by either documented option.
